Add in small optimization for instr comparison

revans2 opened the pull request:
In older versions of Spark contains was not exposed, so we still see some people use tricks like `instr > 0`. This optimizes that operations. It is not a huge win. It just removes a > comparison that is not needed.

This fixes #10578

revans2 commented:
build

revans2 commented:
I should say that I ran some performance numbers, and like you would expect, it really comes down to the amount of time instr is processing the data. So for small strings I saw a 1.6x speedup. For very large strings it went down to just a 1.1x speedup.  Note that this is end to end for a single query that did 20 instr comparisons that could all be converted into contains.

gerashegalov commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:
```diff
@@ -1962,6 +1962,29 @@ case class GpuStringToMap(strExpr: Expression,
+        // stringInstr(A, B) > 0 becomes contains(A, B)
+        GpuContains(str, substr)
+      case GpuGreaterThanOrEqual(GpuLiteral(v, _), GpuStringInstr(str, substr: GpuLiteral))
+        if v == 0 =>
```
```suggestion
        if v == 1 =>
```

gerashegalov commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:
```diff
@@ -1962,6 +1962,29 @@ case class GpuStringToMap(strExpr: Expression,
+object GpuStringInstr {
+  def optimizeContains(cmp: GpuExpression): GpuExpression = {
+    cmp match {
+      case GpuGreaterThan(GpuStringInstr(str, substr: GpuLiteral), GpuLiteral(v, _)) if v == 0 =>
```
nit: here and below we could put value into the pattern, I think it is readable with your comments 
```suggestion
      case GpuGreaterThan(GpuStringInstr(str, substr: GpuLiteral), GpuLiteral(0, _)) =>
```

winningsix commented:
> cessing the data. So for small strings I saw a 1.6x speedup. For very large strings it went down to just a 1.1x speedup. Note that this is end to end for a single query that did 20 instr comparisons that could all be converted into contains.

Just curious. Most benefits come from the elimination of binary operation "greatThan", right?

revans2 commented:
> Just curious. Most benefits come from the elimination of binary operation "greatThan", right?

Yes, that is what I saw. It shifted the time from one kernel to another. We are also going to investigate if there are things we can do to speed up these kernels, especially for long strings.

revans2 commented:
@gerashegalov @winningsix @firestarman I have addressed the review comments. Thanks for picking up on my mistakes. I have fixed them and manually verified that we are doing the right thing in all cases.

revans2 commented:
build

gerashegalov reviewed (approved):
LGTM
